---
title: "Assignment1.rmd"
date: "2023-02-12"
author: "Metyu Melkonyan"
output: 
  html_document:
    toc: yes
  html_notebook:
    toc: yes
---

## Installing dependencies and dowloading the packages
<br>
These dependencies has are already included with the command in the docker file
``` RUN R -e "BiocManager::install(c('DESeq2', 'pheatmap'))" ``` 

```{r eval=TRUE}

if (!requireNamespace("BiocManager", quietly = TRUE)){
  install.packages("BiocManager")}

if (!requireNamespace("GEOmetadb", quietly = TRUE)){
  BiocManager::install("GEOmetadb")}

if (!requireNamespace("GEOmetadb", quietly = TRUE)){
  install.packages("knitr")}

if (!requireNamespace("edgeR", quietly = TRUE)){
  BiocManager::install("edgeR")}

if (!requireNamespace("biomaRt", quietly = TRUE)){
  BiocManager::install("biomaRt")}

if (!requireNamespace("DBI", quietly = TRUE)){
  install.packages("DBI")}

if (!requireNamespace("GEOquery", quietly = TRUE)){
  BiocManager::install("GEOquery")}

```


## Task 1: Selecting an Expression Dataset
I am obtaining the oncogene expression data from the previous query. The following query allows for the **GSE164730** expression data retrieval.

```{r eval=TRUE}
gseacession_number <- 'GSE131222'
supfiles = GEOquery::getGEOSuppFiles(gseacession_number)

fnames = rownames(supfiles)
```
<br>
Reading the gene expression data and understanding if the data is processable.

```{r eval=TRUE}
 b2 = read.delim(fnames[1],header =TRUE)
 head(b2)
```
<br>
```{r eval=TRUE}
gse <- GEOquery::getGEO("GSE131222",GSEMatrix = FALSE)
knitr::kable(data.frame(head(GEOquery::Meta(gse))), format = 'html')
```
<br>
```{r eval=TRUE}
current_gpl <- names(GEOquery::GPLList(gse))[1]
current_gpl_info <- GEOquery::Meta(GEOquery::getGEO(current_gpl))

current_gpl_info$title

current_gpl_info$submission_date

current_gpl_info$last_update_date

current_gpl_info$organism

length(current_gpl_info$series_id)
length(current_gpl_info$sample_id)
```
<br>
Expression data retrieval from the GSE131222
```{r eval=TRUE}
sfiles = GEOquery::getGEOSuppFiles('GSE131222')
fnames = rownames(sfiles)

```
<br>
There is only one supplementray file. I am specifying it on the next query.
```{r eval=TRUE}
GLI2_exp =read.delim(fnames[1],header=TRUE,
check.names = FALSE)

knitr::kable(GLI2_exp[1:15,1:9], format = "html")
```

## Task 2: Data Cleaning and analyzing the quantitative data


<br>
Number of PDCD1 gene for the measurements earlier filtering
```{r eval=TRUE}
dim(GLI2_exp)
colnames(GLI2_exp)
```
<br> 
```{r eval=TRUE}
summarized_gene_counts <- sort(table(GLI2_exp$gene),
                       decreasing = TRUE)

knitr::kable(table(GLI2_exp$gene)[1:10], format="html")

knitr::kable(summarized_gene_counts[which(summarized_gene_counts>1)[1:10]])
```
<br> 
The previous query allows to see if the frequencies of the duplicate data. There are multiple and only 10 of them are retrived.
```{r eval=TRUE}
cpms = edgeR::cpm(GLI2_exp[,3:8])
rownames(cpms) <- GLI2_exp[,2]

keep = rowSums(cpms >1) >= 6
GLI2_exp_filtered = GLI2_exp[keep,]
dim(GLI2_exp_filtered)
colnames(GLI2_exp)
```
<br>
Less duplicate data observed ,after filtering the column names are different such as 2-Mar ! This is not going to cause any issue due to its low frequency
```{r eval=TRUE}
summarized_gene_counts_filtered <- sort(table(GLI2_exp_filtered$gene))

knitr::kable(summarized_gene_counts_filtered[which(summarized_gene_counts_filtered >1)[1:10]],format = "html")
```
<br>
```{r eval=TRUE}
samples <- data.frame(lapply(colnames(GLI2_exp)[3:8],
                FUN=function(x){unlist(strsplit(x, 
                                          split = "\\YR"))[c(2)]}))
colnames(samples) <- colnames(GLI2_exp)[3:8]
rownames(samples) <- c("cell_type")
samples <- data.frame(t(samples))
samples
```

##  Task 3: Normalization by Using edgeR Package



#### Distritibution of The Expression Data


#### Boxplot
```{r eval=TRUE}
data2plot <- log2(edgeR::cpm(GLI2_exp_filtered[,3:8]))
boxplot(data2plot, xlab = "Samples", ylab = "log2 CPM", las = 2, cex = 0.5, cex.lab=0.5,
  cex.axis=0.5, main = "GLI2 RNASeq samples")
abline(h = median(apply(data2plot, 2, median)),
  col = "green", lwd = 0.6, lty = "dashed")
```

#### Density Plot

```{r eval=TRUE}
counts_density <- apply(log2(edgeR::cpm(GLI2_exp_filtered[,3:8])),
                        2, density)
xlim <- 0; ylim <- 0
for (i in 1:length(counts_density)) {
  xlim <- range(c(xlim, counts_density[[i]]$x));
  ylim <- range(c(ylim, counts_density[[i]]$y))
}

cols <- rainbow(length(counts_density))
ltys <- rep(1, length(counts_density))

plot(counts_density[[1]], xlim=xlim, ylim=ylim, type="n",
     ylab="Smoothing density of log2-CPM",
     main="", cex.lab=0.85)
for (i in 1:length(counts_density))
  lines(counts_density[[i]], col=cols[i], lty=ltys[i])

legend("topright", colnames(data2plot),
       col=cols, lty=ltys, cex=0.75,
       border="blue", text.col = "green4",
       merge = TRUE, bg = "gray90")
```
<br>
TMM method is applied to further normalize the values of expression per each cell type group.
```{r eval=TRUE}
filtered_data_matrix <-as.matrix(GLI2_exp_filtered[,3:8])
rownames(filtered_data_matrix) <- GLI2_exp_filtered$gene
d = edgeR::DGEList(counts=filtered_data_matrix, group=samples$cell_type)
```
<br>
Normalization factor is calculated based upon expression and the library size each group has.
```{r eval=TRUE}
d = edgeR::calcNormFactors(d)
normalized_counts <- edgeR::cpm(d)
d$samples[1:6,]
```
<br>
MDS plotted for each group. There is very high distance between samples of EV treatment and GLI2  treatment.
```{r eval=TRUE}
edgeR::plotMDS.DGEList(d, labels=rownames(samples),
  col = c("darkgreen","blue")[factor(samples$cell_type)])
```

## Task 4: Interpretation of Cleaned, Normalized Expression Data


<br>
Setting up the connection with ENSEMBL via biomaRt package.Human ensembl database is connected
```{r eval=TRUE}
ensembl <- biomaRt::useMart("ensembl")
datasets <- biomaRt::listDatasets(ensembl)
knitr::kable(head(datasets),format = "html")
knitr::kable(head(datasets[grep(datasets$dataset,
                pattern = "sapiens"),]), format = "html")
ensembl = biomaRt::useDataset("hsapiens_gene_ensembl",mart=ensembl)
```


#### Filters


```{r eval=TRUE}
biomart_human_filters <- biomaRt::listFilters(ensembl)
knitr::kable(biomart_human_filters[
  grep(biomart_human_filters$name,pattern="hgnc_symbol"),],
      format="html")
```

#### Attributes


HUGO symbols are used. Because there is not ensembl gene id on the data set.No need for conversion requried.
```{r eval=TRUE}
knitr::kable(biomaRt::searchAttributes(mart = ensembl, 'hgnc'),format="html")
```
<br>
The next query is used for validating if the ensembl dataset has the same HUGO symbols as my dataset has. By alligning two data sets, duplicate HUGO symbols or misused names will be removed.
```{r message=FALSE}
validation <- "gli2_id_conversion.rds"
if(file.exists(validation)){
  GLI2_id_conversion <- readRDS(validation)
} else{
 GLI2_id_conversion <- biomaRt::getBM(attributes = c("hgnc_symbol"),
                        filters = c("hgnc_symbol"),
                        values = GLI2_exp_filtered$gene,
                        mart = ensembl)
  saveRDS(GLI2_id_conversion, validation)
}
readRDS(validation)
```
<br>
```{r eval=TRUE}
length(which(rownames(normalized_counts) %in% 
                     GLI2_id_conversion$hgnc_symbol))
```
<br>
The difference of our HUGO symbols of the data frame we were using and the data frame we retrieved from biomaRt
```{r eval=TRUE}
nrow(normalized_counts)-length(which(rownames(normalized_counts) %in%
    GLI2_id_conversion$hgnc_symbol))
```
<br>
Merging the the normalized data frame with the converted data frame. The gene symbols are sorted and normalized values (first 5 rows) are respective to the treatments made

```{r eval=TRUE}
normalized_counts_annot <- merge(GLI2_id_conversion,normalized_counts,
by.x = 1, by.y = 0, all.y=TRUE)
knitr::kable(normalized_counts_annot[1:5,1:7],type = "html")
```
<br>
No missing HUGO symbol found
```{r eval=TRUE}
hgnc_symbol_missing_gene <- normalized_counts_annot$hgnc_symbol[
 which(is.na(normalized_counts_annot$hgnc_symbol))]
length(hgnc_symbol_missing_gene)
nrow(normalized_counts_annot)
```

# Interpretation
<br>

### What are the control and test conditions of the dataset?

The YAPC-iEV (YRE1A,YRE1B,YRE2A) cells are control in the dataset. They are compared against the YAPC-iGLI2(YRG1A,YRGA,YRG2B) to see the RNA expression via gene set enrichment analysis(GSEA) on Figure 3.D

### Why is the dataset of interest to you? 

The dataset important aspect of basal-like subtype switching by GLI2 treatment on basal-like cell lines.PDA is seen among basal-like cells and is important for understanding tumor epithelial-to-mesenchymal(EMT) transition among pancreas cancer cell lineage.

<br>

### Were there expression values that were not unique for specific genes? How did you handle these?

Yes,there were expression values for genes specified (Line 156) I have filtered them. Because there were significant amount of the duplicate genes which might disturb the later analysis.
 
<br>

### Were there expression values that could not be mapped to current HUGO symbols?

The expression values were unique to each gene with respective to the treatments.
The dataset already has HUGO symbols. This is why I did not convert them to ENSEMBL gene ID's. However for comparison and data validty I have also converted and alligned them with the ones in ENSEMBL dataset contains. I also compared that they have don't have a lot of difference (Line299)
<br>

### How many outliers were removed?

I have removed 15497 outliers having low counts.
<br>

### How did you handle replicates?
I had to do seperate research on the replicated genes. Most of these ones are non-coding RNA  such as micro-RNA and so on for example (https://www.ncbi.nlm.nih.gov/gene?Db=gene&Cmd=DetailsSearch&Term=102465753)
<br>

### What is the final coverage of your dataset?
The final data set includes 11527 genes with respective treatments made grouped,cleaned and normalized values.
<br>

# References

Adams, C. R., Htwe, H. H., Marsh, T., Wang, A. L., Montoya, M. L., Subbaraj, L., Tward, A. D., Bardeesy, N., & Perera, R. M. (2019). Transcriptional control of subtype switching ensures adaptation and growth of pancreatic cancer. ELife, 8. https://doi.org/10.7554/elife.45313

Bioconductor - home. (n.d.). Bioconductor.org. Retrieved February 13, 2023, from https://www.bioconductor.org/

EdgeR. (n.d.). Bioconductor. Retrieved February 13, 2023, from https://bioconductor.org/packages/release/bioc/html/edgeR.html

Ensembl genome browser 109. (n.d.). Ensembl.org. Retrieved February 12, 2023, from http://useast.ensembl.org/index.html

GEO overview. (n.d.). Nih.gov. Retrieved February 13, 2023, from https://www.ncbi.nlm.nih.gov/geo/info/overview.html

National center for biotechnology information. (n.d.). Nih.gov. Retrieved February 12, 2023, from https://www.ncbi.nlm.nih.gov/

Xie, Y., Allaire, J. J., & Grolemund, G. (2018). R markdown: The definitive guide. CRC Press.


